enhancements/update/upgrades-blocking-on-ack.md

```diff
@@ -0,0 +1,211 @@
+```
+CVO will set `Upgradeable=false` which in turn blocks cluster upgrades to 4.9 until the administrator has added the key/value pair `ack-4.8-kube-122-api-removals-in-4.9: "true"` to the `admin-acks` configmap as shown below.
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  annotations:
+    release.openshift.io/create-only="true"
+  name: admin-acks
+  namespace: openshift-cluster-version
+data:
+    ack-4.8-kube-122-api-removals-in-4.9: "true"
+```
+Until this is done the `Conditions` console page will display an `Upgradeable` condition message directing the administrator to the relevant OCP documentation.
+
+If the administrator wants to upgrade to 4.9 without performing the acknowledgment `force` can be used.
+
+### User Stories
+
+* As a cluster administrator, I want to be warned that APIs will be removed if I upgrade to 4.9 before I upgrade.
+
+* As a cluster administrator, I want to be instructed on how to check my cluster for potential impacts of API removal.
+
+* As a cluster administrator, I want to be able to override an upgrade block and upgrade anyway.
+
+* As an OCP developer I want a way to introduce additional "acknowledgment based" upgrade blocks for blocking minor level upgrades.
+
+* As an OCP product team we do not want customers to be surprised by workloads breaking due to APIs being removed after they upgrade.
+
+### Implementation Details/Notes/Constraints
+
+### Risks and Mitigations
+
+* A common risk is human error resulting in an incorrect key/value pair being placed in the `admin-acks` configmap.
+This type of error will not change `Upgradeable=false` and therefore the administrator will know a mistake has been made since the `Upgradeable` condition message will continue to be displayed.
+CVO will log an error whenever an unrecognized key is found in the `admin-acks` configmap.
+* Another risk is that the `admin-acks` configmap gets removed.
+This will result in `Upgradeable=false` and the standard `Upgradeable` condition message will be displayed with the additional message that the configmap cannot be found and should be created.
+CVO will recreate the `admin-gates` configmap if removed since it will be managed by CVO.
+CVO will log an error whenever the `admin-acks` configmap has been removed.
+* Unable to access the `admin-acks` configmap.
+This will result in `Upgradeable=false` and the standard `Upgradeable` condition message will be displayed with the additional message that the configmap cannot be accessed.
+CVO will log an error whenever there is an issue accessing the `admin-acks` configmap.
+* Unable to access the `admin-gates` configmap.
+Since CVO will not be able to determine what if any minor level upgrade blocks are defined CVO will set `Upgradeable=false` and set the displayed `Upgradeable` condition message to indicate that the `admin-gates` configmap cannot be accessed.
+CVO will log an error whenever there is an issue accessing the `admin-gates` configmap.
+
```

Comment:
Is there a risk for EUS updates if only some of the acks are provided, allowing a partial upgrade stopping at 4.x+1 but not progressing to 4.x+2? Or does the logic check for all required acks between 4.x and the target version?

Reply:
there is no automated update from one EUS to the next, it's already a manual process where the admin does each sequential update explicitly, so while they might update from 4.x to 4.x+1 and then find they can't progress to 4.x+2 w/o doing some additional ack(which i agree isn't great if they suddenly find they need a week to resolve the ack before they can continue their upgrade process), it's not like it's blocking an automated "get me from 4.x to 4.x+3" process the admin kicked off(because no such process exists today, afaik).

i'm not really sure how we'd solve this.  
1) when you're at 4.6, 4.6 does not know what requirements may exist to get from 4.8 to 4.9, unless we start backporting stuff all the way to 4.6 once we discover it in 4.8/4.9.

2) when you're on 4.6 we don't actually know your intention as far as what level you are intending to get to next (as i said, there's no "upgrade me to 4.10" button, it's just "upgrade me to 4.7.  and then upgrade me to 4.8.  and then.....", so it seems difficult to communicate "hey, fyi, you're on 4.6 and that's fine, and you can get to 4.7 and 4.8, that's fine too, but if you intend to get to 4.9 you're going to get stuck at some point".  We might have a hint based on you being on the "EUS" channel instead of the "stable" channel, though i'm not clear what our longterm vision is for that channel(are we going to put 4.7+4.8+4.9 into that channel when 4.10EUS comes out?  or are we going to tell people to switch to the stable channel to get to 4.10 and then switch back to the EUS channel?).

it feels like something that belongs in the EUS upgrade guidance(hopefully we have such a thing), i'm sure this isn't the only "gotcha" we want admins to think about when they are getting ready to go from 4.6 to 4.10, as opposed to just going from 4.6 to 4.7.

Reply:
Given the CVO would be updated at each step along that path I think there's no real concern here.

We've sort of sputtered out on ensuring that 4.N can signal all of the requirements to get to 4.N+4 and I don't see us achieving that goal so I'm not too worried about it. If we stick with 4 minor stride between EUS versions maybe we try to address this in 4.10 to 4.14, which means we'd need to revisit this as soon as we start 4.10 dev.

Reply:
OK, it sounds like we can address my concerns via documentation. It would be good to include it as a risk in the design, though.